### DISCUSSION TITLE
Shortcodes for variable content based on the output format

### DISCUSSION THREAD

**Discussion by jmbuhr** [quarto-dev/quarto-cli#373] [Feature Requests]
I am currently working on a paper in which I am including videos (which works great with the regular markdown image syntax btw.). However, this of course does not work for pdf output. For my Master's Thesis I solved this with Rmarkdown using R code:

````
```{r, results='asis'}
if (is_html_output()) {
  cat("
> **Dear reader**: You are holding in your hands the digital
  version of this document.
")
}
if (is_latex_output()) {
  cat("
> **Dear reader**: You are holding in your hands the
  print version of this document.
")
}
```
````
(https://github.com/jmbuhr/hdReports/blob/c55392dcf19cab8d1fd84796a83214fa848dc4ce/inst/rmarkdown/templates/hdReport/skeleton/skeleton.Rmd#L54-L67)

but for quarto it feels like there would be a more elegant language agnostic solution straight in quarto with some sort of shortcode. Is this feasible?

**Comment by cderv** [comment]
I believe it is on the roadmap. Shortcodes are already a thing for Quarto (https://quarto.org/docs/authoring/shortcodes-and-filters.html#shortcodes) and additional shortcodes for inserting videos, tweets, or others are definitely something that should be done ! 

Maybe we could use this discussion to identify the shortcodes that would be useful ? 

I am thinking  (mainly for HTML 😅 ) : 
* `{{< video url attributes >}}`, and maybe more specific `{{< youtube id >}}`
 * `{{< tweet id >}}`
 * `{{< gist id >}}`

For custom shortcodes, it may be something feasible using a Lua Filters (that would output HTML or LaTeX syntax depending on Pandoc's output format.

**Reply by jmbuhr** [reply]
I wonder how far we can or want to tap into pandocs already existing shortcodes that are being used in template. For example pandocs `$if(...)$...$endif$`. As far as I know those can only be used in the actual templates and not in the document content, event though they would already solve some usecases.

**Reply by cderv** [reply]
I see this syntax as rather as templating syntax than a shortcode syntax. This kind of logical syntax is used for custom listing for example with ejs template (https://quarto.org/docs/websites/website-listings-custom.html#overview)

but I see the point of having a language agnostic syntax to include content conditionally and not using a code chunk like you did with R. 
I wonder how Divs and Spans + other component of Pandoc are enough to add this logic with a lua filter . [Adding filter is quite easy](https://quarto.org/docs/authoring/shortcodes-and-filters.html#using-filters), but writing them is not straightforward, especially for simple use case as your example. 

Quarto supports already variable, maybe there could be a way to add conditional usage of variable depending on format. Like using `digital` or `print` as in your example, but with a simple syntax. 🤔

**Reply by jmbuhr** [reply]
Wrote my first eve lua filter and it works :) see https://github.com/quarto-dev/quarto-cli/pull/386

**Reply by ericpgreen** [reply]
@jmbuhr I have a [very similar use case](https://community.rstudio.com/t/conditionally-exclude-some-bookdown-content/113844/2?u=ericpgreen). In rmd I was using `is_latex_output()` and `is_html_output()` in the chunk options along with `cat(tufte::margin_note())` to conditionally print to the margin. Specifically, when the format is html I do things like add links and embed videos and gifs. When it's latex, I include only text plus small QR code images that point the reader to multimedia content.

I'm transitioning to Quarto and failing to get this to work:

```
:::web-only {.column-margin}
Test 1
:::

:::print-only {.column-margin}
Test 2
:::
```

I think this is because your PR is still open, right? @cderv do you think it will be possible to combine this approach with `{.column-margin}`?

```
> sessionInfo()
R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Catalina 10.15.7

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods  
[7] base     

loaded via a namespace (and not attached):
 [1] compiler_4.0.2  here_1.0.1      fastmap_1.1.0  
 [4] tufte_0.12      rprojroot_2.0.3 htmltools_0.5.2
 [7] tools_4.0.2     rmarkdown_2.12  knitr_1.37     
[10] xfun_0.30       digest_0.6.29   rlang_1.0.2    
[13] renv_0.14.0     evaluate_0.15 
```

**Reply by jmbuhr** [reply]
Until there is an official solution you can use my lua filter by downloading it and including it in the the yaml frontmatter:
https://github.com/hits-mbm-dev/paper-talin-loop/blob/1d85e2b9d3c7b76b38ed185182abfa45b18eedaa/index.qmd#L57-L59
https://github.com/hits-mbm-dev/paper-talin-loop/blob/1d85e2b9d3c7b76b38ed185182abfa45b18eedaa/www/conditional-content.lua

**Reply by ericpgreen** [reply]
Thanks, @jmbuhr. I appreciate that. I grabbed the file and added it to the yaml. It works unless I add `{.column-margin}`. 

```
:::web-only {.column-margin}
Test
:::
```

I think I need to use the `.column-margin` class to put content in the margin, and I assumed it might come after the filter name. This is quarto day 1 for me, so I might not be understanding something about how it works. Any ideas?

```
::: {.column-margin}
![A margin image](image.png)
:::
```

Your paper repository is amazing, btw. I'll be coming back to it for general learning.

**Reply by jmbuhr** [reply]
Thank you for your kind words!

`:::web-only` is a shorthand notation for a div with a single class, so it gets translated to `:::{.web-only}` (mind the dot). This means if you want a div with multiple classes they all need to go inside of the curly braces: `:::{.web-ony .column-margin}`.
Since quarto is based on pandoc, the pandoc manual can sometimes come in handy as well: https://pandoc.org/MANUAL.html#extension-fenced_divs and https://pandoc.org/MANUAL.html#extension-fenced_code_attributes

**Reply by ericpgreen** [reply]
This really helped me, @jmbuhr. Thanks. I didn't know it was shorthand for the dot version.

**Comment by cderv** [comment]
For cross reference, related discussion in https://github.com/quarto-dev/quarto-cli/discussions/676 also following this issue should help know when something will be available: https://github.com/quarto-dev/quarto-cli/issues/646

**Comment by cderv** [comment]
See the new feature as answered in https://github.com/quarto-dev/quarto-cli/issues/646#issuecomment-1115459837

````markdown
::: {.content-hidden when-format="pdf"}
This will not appear in a PDF
:::

::: {.content-visible unless-format="docx"}
This will not appear in word output
:::

::: {.content-hidden unless-format="html"}
This will only be visible in HTML output
:::
````

This should help regarding the discussion here. Please give us feedback to know if that answer all or still missing elements. 

Thanks !

**Reply by ericpgreen** [reply]
@cderv you mentioned shortcodes for embedding video urls in an earlier reply:

> {{< video url attributes >}}, and maybe more specific {{< youtube id >}}

Do these shortcodes exist yet? I don't think so. 

Here's my current approach with the new `content-hidden` shortcodes your team created (which are great!). I'm trying to get the video to appear in the margin:

```
::: {.content-hidden unless-format="html" .column-margin}
![](https://youtu.be/491LFqvofxo?t=492)
:::
```

I get a missing image placeholder.  I tried the following but the video appears to overflow the margin (I tried with several smaller sizes, too):

````
::: {.content-hidden unless-format="html" .column-margin}
```{r}
#| label: fig-myfig
#| results='asis'
#| eval=knitr::is_html_output()
cat("<iframe width='300' height='169' src='https://www.youtube.com/embed/491LFqvofxo?start=492' frameborder='0' allow='accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture' allowfullscreen></iframe>")
:::
````

**Reply by jmbuhr** [reply]
No need for R code to embedd raw html here. Try this:
```
<style>
iframe {
    width: 100%;
}
</style>

:::: {.content-visible when-format="html"}
:::{#fig-pandas}
<iframe height="500px" src="https://www.youtube.com/embed/XdM6c4juY1g" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Look at these pandas!
:::
::::
```

**Reply by jmbuhr** [reply]
(The iframe styling prevents the overflowing of the video, the rest is just there to show that you can even put it in a quarto figure and thus reference it later)

**Reply by cderv** [reply]
@ericpgreen no those shortcodes does not exist yet but there is now a way to create your own shortcode https://quarto.org/docs/authoring/shortcodes.html#custom-shortcodes

> I'm trying to get the video to appear in the margin:

@jmbuhr I think @ericpgreen wants video in the margin using `:::: {.column-margin}`. I think this does not play way currently with the `.content-visible` feature. We can probably try to make it work together. You'll have to tweak the HTML though to have a video in the margin in correct size.

**Reply by jmbuhr** [reply]
Ohh, my bad, I read it as the opposite. Probably because that is the problem people encounter more often.

**Reply by cderv** [reply]
@ericpgreen I think you need to use nested divs for using both `content-hidden` or `contetn-visible` and `column-margin` together. 
You should be able to mix with @jmbuhr solution which seems good : I don't see either why you would need R code chunk in this case. 

````markdown
:::::{.content-hidden unless-format="html"}

:::: {.column-margin}

:::{#fig-pandas}
<iframe height="500px" src="https://www.youtube.com/embed/XdM6c4juY1g" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Look at these pandas!
:::

::::

:::::
````

**Reply by ericpgreen** [reply]
Thanks @jmbuhr and @cderv. Confirming that the nested `content-hidden` and `column-margin` approach works!

<img width="1313" alt="Screen Shot 2022-05-10 at 12 41 38 PM" src="https://user-images.githubusercontent.com/546433/167679959-40db4ad5-0223-40a9-ba02-3a070c372e4b.png">

**Reply by cderv** [reply]
@ericpgreen This was a bug that should now be fixed in 8d40ba3d0252259bd86b0c5442f5cf1fd578abef 
it should be available in the next build in https://github.com/quarto-dev/quarto-cli/releases

So you could use one div 

````markdown
::::{.column-margin .content-hidden unless-format="html"}
:::{#fig-pandas}
<iframe height="500px" src="https://www.youtube.com/embed/XdM6c4juY1g" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Look at these pandas!
:::
::::
````

Thanks for the report

**Reply by ericpgreen** [reply]
@cderv should the latest build also allow combining the following?

`::::{.column-margin .content-hidden unless-format="pdf"}`

I updated to the development version today and the above is failing for me (i.e., content appears but not in margin), but the nested approach works.

**Reply by cderv** [reply]
Yes this above https://github.com/quarto-dev/quarto-cli/discussions/373#discussioncomment-2724275 works to include a iframe in the margin. Can you confirm this works ? 

In your example, you are trying to include content in margin for a PDF. I don't know what it is but it should be LaTeX compatible format otherwise it may be ignored. Can you share full minimal reprex of what is not working ?

**Reply by ericpgreen** [reply]
Thanks @cderv. I created a new repo and a reprex of the issue: https://github.com/ericpgreen/quarto_example/blob/main/index.qmd. 

Only the nested version is working for PDF and html (it was working for html in an earlier version). I installed the development version and updated to 393 just prior to posting this.

It's not a problem to use nested divs if that's required. I'm about to convert a large project, so my main goal is to figure out the best way to do it currently.

**Reply by cderv** [reply]
We'll look closer to this, and we may indeed recommend nested divs as it will make more clear in layout what is happening. We'll keep you updated

**Comment by jjallaire** [comment]
Complete docs here: https://quarto.org/docs/authoring/conditional.html

**Comment by mcanouil** [comment]
Closing this as completed.
